### PR TITLE
Korjataan laskugeneroinnin suorituskyky, ja parannellaan koodia

### PR DESCRIPTION
**Pull request by Gekkio**

### CODE REVIEW

File: service/src/main/kotlin/fi/espoo/evaka/invoicing/service/InvoiceGenerator.kt [hunk 1]
```diff
@@ -126,7 +120,10 @@ class InvoiceGenerator(
                 .toSet() +
                 permanentPlacements.keys +
                 temporaryPlacements.values.flatMap { pairs -> pairs.map { it.second.child.id } }
-        val operationalDaysByChild = tx.getOperationalDatesForChildren(range, allChildren)
+        val operationalDaysByChild =
+            tx.getOperationalDatesForChildren(range, allChildren).mapValues {
+                DateSet.of(it.value.map(LocalDate::toFiniteDateRange))
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/invoicing/service/InvoiceGenerator.kt:125]: might be more readable to add an overload for DateSet.of which accepts LocalDates directly

File: service/src/main/kotlin/fi/espoo/evaka/invoicing/service/DraftInvoiceGenerator.kt [hunk 2]
```diff
@@ -113,7 +112,13 @@ class DraftInvoiceGenerator(
                         data.areaIds,
                         data.businessDays,
                         data.feeThresholds,
-                        absencesByChild,
+                        getChildAbsences =
+                            memoize { child: ChildId ->
+                                (absencesByChild[child] ?: emptyList()).associateBy(
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/invoicing/service/DraftInvoiceGenerator.kt:117]: I think there can be up to 2 absence types per date so `associate` kind of randomly looses one? Not sure if the same issue exists in the original code too.

Comment (Gekkio) [reply service/src/main/kotlin/fi/espoo/evaka/invoicing/service/DraftInvoiceGenerator.kt:117]: Only billable absences are fetched so there can be only one per date. It's just not obvious from the type `List<AbsenceStub>`, which is one of the reasons why I've changed most functions to accept absence data using types that either make this either explicit (e.g. `Map<LocalDate, AbsenceType>`) or remove the need for this detail completely (e.g. `hasPlannedAbsence: (date: LocalDate) => Boolean`)

Comment (Joosakur) [reply service/src/main/kotlin/fi/espoo/evaka/invoicing/service/DraftInvoiceGenerator.kt:117]: ah gotcha, all good then

**Review by Joosakur — APPROVED** [review]

**Pull request merged**